### PR TITLE
feat: implement `allow_email_share` field 📬

### LINKED ISSUE #51: Share member contact info with Chapter Leaders 📬
### Description

✅ It would be great to implement something that gives permission for ColorStack to share a Member's Email with local ColorStack Chapter leaders.

### Problem / Value Proposition

📩 Chapter Leaders (like myself) are always looking for ways to engage with the community and to reach as many people as possible. In my case there are hundreds if not upwards of a thousand ColorStack members across the NYC metro area. But Johana can't just give me their contact info for legal reasons. Implementing this will help bridge the gap between chapter leaders, and the community. 

### Additional Context

💻 This will likely be implemented as a checkbox somewhere in this folder [here](apps/member-profile/app/routes) my guess is either the [emails](apps/member-profile/app/routes/_profile.profile.emails.tsx) page or the [general](apps/member-profile/app/routes/_profile.profile.general.tsx) page

- You can reference how we built the component for the [BirthdateNotification](https://github.com/colorstackorg/oyster/blob/e3ca5dd64ffb53cb75e139c374a3c146d78a2d3c/apps/member-profile/app/shared/components/profile.personal.tsx#L8)
- The checkbox will store a boolean value in the database, and it's default should be True.
- This will require a Database migration similar to [this](packages/core/src/infrastructure/database/migrations/2024-03-18T19:45:26Z-birthday-default.ts) one. 

Good and luck and be sure to see our [contributing](https://github.com/colorstackorg/oyster/blob/main/CONTRIBUTING.md) guide!👋

### PR DESCRIPTION
**Pull request by tomas-salgado**
## Description ✏️
Implement a checkbox on the emails page that allows members to opt-in or opt-out of having their email shared with Chapter Leaders. 

Closes #51 

- adds `allow_email_share` field to `students` table in db
- default gives `allow_email_share` a value of `True`
- implements db migrations
- implement frontend checkbox on emails page, where initial value (checked or not) is in line with `allow_email_share` value in database for that user
- implement backend functionality such that database is updated when checkbox is clicked

After:
Screenshot 2024-04-02 at 12 22 26 PM

Before:
Screenshot 2024-04-01 at 10 36 04 PM

## Type of Change 🐞

- [X] Feature - A non-breaking change which adds functionality.
- [ ] Fix - A non-breaking change which fixes an issue.
- [ ] Refactor - A change that neither fixes a bug nor adds a feature.
- [ ] Documentation - A change only to in-code or markdown documentation.
- [ ] Tests - A change that adds missing unit/integration tests.
- [ ] Chore - A change that is likely none of the above.

## Checklist ✅

- [x] I have done a self-review of my code.
- [x] I have manually tested my code (if applicable).
- [x] I have added/updated any relevant documentation (if applicable).

### CODE REVIEW

File: packages/core/src/infrastructure/database/migrations/2024-03-31T15:39:09Z-share-email-default.ts
Comment (tomas-salgado) [inline packages/core/src/infrastructure/database/migrations/2024-03-31T15:39:09Z-share-email-default.ts:36]: Does it make more sense to combine share-email-default.ts with share-email.ts rather than have two separate migration files? Here I was following the example as seen with [birthday-notification](https://github.com/colorstackorg/oyster/blob/main/packages/core/src/infrastructure/database/migrations/2024-03-18T17%3A06%3A13Z-birthday-notification.ts) and [birthday-default.](https://github.com/colorstackorg/oyster/blob/main/packages/core/src/infrastructure/database/migrations/2024-03-18T19%3A45%3A26Z-birthday-default.ts). Not sure if the separate files are necessary

Comment (ramiAbdou) [reply packages/core/src/infrastructure/database/migrations/2024-03-31T15:39:09Z-share-email-default.ts:36]: Great question, yes please combine them! We only had to do that for the `birthday-notification`/`birthday-default` because we forgot to ship it with a default value, and so we needed to add it after.

File: packages/core/src/infrastructure/database/migrations/2024-03-31T15:36:45Z-share-email.ts
Comment (ramiAbdou) [inline packages/core/src/infrastructure/database/migrations/2024-03-31T15:36:45Z-share-email.ts:6]: I _generally_ like the idea of having a prefix that hints that a field is a `boolean`. What do you think of `allow_share_email` or something of that sort?

Comment (tomas-salgado) [reply packages/core/src/infrastructure/database/migrations/2024-03-31T15:36:45Z-share-email.ts:6]: Ah yeah that's definitely more clear. How about `allow_email_share`?

Comment (ramiAbdou) [reply packages/core/src/infrastructure/database/migrations/2024-03-31T15:36:45Z-share-email.ts:6]: Even better, let's go with that!

File: packages/core/src/infrastructure/database/migrations/2024-03-31T21:37:51Z-share-email.ts
Comment (ramiAbdou) [inline packages/core/src/infrastructure/database/migrations/2024-03-31T21:37:51Z-share-email.ts:6]: Let's combine all of the previous statements into one!

```
await db.schema
  .alterTable('students')
  .addColumn('allow_email_share', 'boolean', (column) => {
    return column.notNull().defaultTo(true);
  })
```

Sorry for the confusion, I know the other 2 migration files probably didn't make it clear that we could do this!

File: packages/core/src/infrastructure/database/migrations/2024-03-31T21:37:51Z-share-email.ts
Comment (tomas-salgado) [inline packages/core/src/infrastructure/database/migrations/2024-03-31T21:37:51Z-share-email.ts:16]: I believe the down function can also be simplified to only this, because when we drop the column, this will remove the default and not null requirements. Please let me know if I'm missing something!

File: apps/member-profile/app/shared/components/profile.personal.tsx
Comment (tomas-salgado) [inline apps/member-profile/app/shared/components/profile.personal.tsx:17]: I think this label could be better/clearer. I was thinking "Share my email with Chapter Leaders to connect with local ColorStack communities!", but this ends up being a bit long

File: apps/member-profile/app/shared/components/profile.personal.tsx
Comment (tomas-salgado) [inline apps/member-profile/app/shared/components/profile.personal.tsx:17]: I wasn't sure about the label for this. I was also thinking "Share my email with Chapter Leaders to connect with local ColorStack communities!" but this seems long.

File: apps/member-profile/app/shared/components/profile.personal.tsx
Comment (tomas-salgado) [inline apps/member-profile/app/shared/components/profile.personal.tsx:8]: I put this function in this file because it imitates the style of the other checkboxes in this file, but I also wasn't sure because technically this checkbox is implemented in `_profile.profile.emails.tsx` not `_profile.profile.personal.tsx`. There isn't a file `profile.emails.tsx` yet, would this be worth creating just for this function?

File: apps/member-profile/app/routes/_profile.profile.emails.tsx
Comment (ramiAbdou) [inline apps/member-profile/app/routes/_profile.profile.emails.tsx:186]: In this particular case (a 1-field form with checkbox), I think it might more sense if we omitted any "Save" button altogether. What if we just save the update as soon as the checkbox is checked?

To do that, we can do something like:

```
const submit = useSubmit();

<RemixForm onChange={(e) => submit(e.currentTarget)}>
  // ...
</RemixForm>
```

File: apps/member-profile/app/shared/components/profile.personal.tsx
Comment (ramiAbdou) [reply apps/member-profile/app/shared/components/profile.personal.tsx:8]: Honestly, I'm cool not abstracting this component into a separate file. I'm not completely sold on that pattern yet, aside from reusable components.

Feel free to just include this directly in the form, ie:

```tsx
<RemixForm>
  <Form.Field ...>
</RemixForm>
```

File: apps/member-profile/app/shared/components/profile.personal.tsx
Comment (ramiAbdou) [reply apps/member-profile/app/shared/components/profile.personal.tsx:17]: What if we did something like this:

```
<Form.Field description="If you go to school where there is a ColorStack chapter, this will allow that chapter leader to reach out to you about local events and opportunities." error={error} label="Would you like to share your email with chapter leaders?">
  <Checkbox
      defaultChecked={defaultValue}
      label="Share my email with chapter leaders! 🌟 "
      name={name}
       value="1"
  />
</Form.Field>
```

It's a bit more verbose, but the checkbox's description itself isn't too long. Thoughts?

File: apps/member-profile/app/shared/queries/index.ts
Comment (ramiAbdou) [inline apps/member-profile/app/shared/queries/index.ts:222]: In an effort to colocate core backend logic, let's move this to the `core` package and create a use-case in the `member` module. You can find an example [here](https://github.com/colorstackorg/oyster/blob/main/packages/core/src/modules/member/use-cases/join-member-directory.ts).

File: packages/core/src/infrastructure/database/migrations/2024-03-31T21:37:51Z-share-email.ts
Comment (ramiAbdou) [reply packages/core/src/infrastructure/database/migrations/2024-03-31T21:37:51Z-share-email.ts:16]: You got it right! 🔥

**Review by ramiAbdou — CHANGES_REQUESTED** [review]
This is excellent work, @tomas-salgado! I can't believe you ramped up so quickly with this. 😅 Just have a few small suggestions, and we should be good to go!

File: apps/member-profile/app/shared/components/profile.personal.tsx
Comment (tomas-salgado) [reply apps/member-profile/app/shared/components/profile.personal.tsx:17]: Yeah, that's great! With this additional description I thought it might be better to move the button `Add Email`, rather than have it in between the description and `Save`. Thoughts on this look? 
Screenshot 2024-04-02 at 12 41 32 AM

Comment (xwalton) [reply apps/member-profile/app/shared/components/profile.personal.tsx:17]: Think it sounds great. For brevity, perhaps

```
<Form.Field description="If your school has a ColorStack chapter, would you like to stay connected for local events and opportunities?" error={error} label="Would you like to share your email with chapter leaders?">
  <Checkbox
      defaultChecked={defaultValue}
      label="Share my email with chapter leaders! 🌟 "
      name={name}
       value="1"
  />
</Form.Field>
```

Comment (tomas-salgado) [reply apps/member-profile/app/shared/components/profile.personal.tsx]: I was thinking potentially we would want this to allow connections beyond just your school, but all schools in your area (especially useful in NYC for example). So maybe the description could be

_Would you like to share your email with chapter leaders? 
If your local schools have a ColorStack chapter, this will allow that chapter leader to reach out about local events and opportunities. 

Share my email with chapter leaders!_

Comment (xwalton) [reply apps/member-profile/app/shared/components/profile.personal.tsx]: That sounds and looks great so far

Comment (ramiAbdou) [reply apps/member-profile/app/shared/components/profile.personal.tsx]: I think for now, we're only going to share emails with their own school. After we merge this, I can follow-up with Johana if we want to allow all local schools though!

Comment (tomas-salgado) [reply apps/member-profile/app/shared/components/profile.personal.tsx]: Sounds good! I updated the description.

File: apps/member-profile/app/routes/_profile.profile.emails.tsx
Comment (tomas-salgado) [reply apps/member-profile/app/routes/_profile.profile.emails.tsx]: Ok, got it working without the "Save" button. It was a simple change after all 😄

**Review by ramiAbdou — COMMENTED** [review]
I just tested this locally and it works beautifully! 🙌

I pushed a small fix which adds the `id` to the checkbox which makes the label of the checkbox clickable - if you can just merge those changes in, we should be good to ship!

**Review by ramiAbdou — APPROVED** [review]
LGTM! 🚀

**Pull request merged**
